### DISCUSSION TITLE
Adding and removing child nodes in update callback

### DISCUSSION THREAD

**Discussion by keince** [openscenegraph/OpenSceneGraph#1256] [Q&A] [answered]
I need to add and remove nodes dynamically during runtime. 

My plan is to attach an update callback to the group node to facilitate this.

Is it safe or appropriate to call addChild() and removeChild() to add/remove these nodes to the group node in the update callback?

Feedback would be appreciated...

**Comment by robertosfield** [comment, marked as answer]
If you add and  removing children during traversals, including the update traversal, you have to be very careful about avoiding invalidation of the iterators that are being used during traversal.

It is generally safe to add/remove children that are below the node that has the callback , but going back to the node parent's or any node in the parent chain you will invalidate iterators, so those iterators will likely cause a crash as they'll be pointing to now deleted or repurposed memory.

Personally I'd recommend doing add/removing at the top level rather than inside the scene graph.

**Reply by keince** [reply]
Thank you for the feedback Robert. 

Can you clarify what you mean by "add/removing at the top level rather than inside the scene graph"? 

Does this mean that it's best to add/remove nodes before or after viewer.frame() in the while(!viewer.done()) loop?

**Reply by robertosfield** [reply]
I mean at the viewer frame level.  You can implement your own frame loop and put invoke your code before/after the Viewer::updateTraversals() call, or override the Viewer::updateTraversals() or implement an UpdateOperation.

**Reply by keince** [reply]
Thank you. This helps a lot.
